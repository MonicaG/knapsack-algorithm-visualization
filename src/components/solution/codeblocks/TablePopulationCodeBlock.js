import { Light as SyntaxHighlighter } from 'react-syntax-highlighter';
import js from 'react-syntax-highlighter/dist/cjs/languages/hljs/javascript';
import {docco} from 'react-syntax-highlighter/dist/cjs/styles/hljs'; 
import {HIGHLIGHT_COLOUR} from './CodeBlocksCommon'

SyntaxHighlighter.registerLanguage('javascript', js);

function TablePopulationCodeBlock({ item, capacity, index, solutionTable }) {

  function getIfComment(itemFitsInKnapsack) {
    return !itemFitsInKnapsack ? "" : 
`
/* 
T[${index}][${capacity}] = Math.max(T[${index - 1}][${capacity}], (${item.value} + T[${index - 1}][${capacity - item.weight}])) 
T[${index}][${capacity}] = Math.max(${solutionTable[index - 1][capacity]}, ${item.value + solutionTable[index - 1][capacity - item.weight]})
T[${index}][${capacity}] = ${solutionTable[index][capacity]}
*/`
  }

  function getElseComment(itemFitsInKnapsack) {
    return itemFitsInKnapsack ? "" :
`
/* 
T[${index}][${capacity}] = T[${index - 1}][${capacity}]
T[${index}][${capacity}] = ${solutionTable[index][capacity]}
*/`    
  }

  function getCode() {
    return (
`// w: item weight (${item.weight})
// value: item value (${item.value})
// c: current capacity (${capacity})
// i: index (${index})
if (w <= c) { // ${item.weight} <= ${capacity}
  T[i][c] = Math.max(T[i-1][c], (value + T[i - 1][c - w])) ${getIfComment(itemFitsInKnapsack)}
}else {
  T[i][c] = T[i-1][c] ${getElseComment(itemFitsInKnapsack)}
}`
    )
  }

  const itemFitsInKnapsack = item.weight <= capacity
  const itemFitsLineNum = [5,6,7,8,9,10,11]
  const itemDoesNotFitLineNum = [7,8,9,10,11,12]
  return (
    <SyntaxHighlighter 
     language="javascript" 
      style={docco}
      showLineNumbers={true}
      wrapLines={true}
      lineProps={lineNumber => {
                let style = { display: 'block' };
                if (itemFitsInKnapsack && itemFitsLineNum.includes(lineNumber)) {
                  style.backgroundColor = HIGHLIGHT_COLOUR
                }else if (!itemFitsInKnapsack && itemDoesNotFitLineNum.includes(lineNumber)) {
                  style.backgroundColor = HIGHLIGHT_COLOUR
                }
      
                return { style };
              }}
    >
      {getCode(itemFitsInKnapsack)}
    </SyntaxHighlighter>
  )
}
export default TablePopulationCodeBlock;