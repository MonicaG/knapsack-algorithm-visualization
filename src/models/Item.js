import { nanoid } from 'nanoid'

class Item {

  constructor(name, value, weight) {
    this.id = nanoid();
    this.name = name;
    this.value = value;
    this.weight = weight
  }

  toString() {
    return `Item  id: ${this.id} name: ${this.name} value: ${this.value} weight: ${this.weight}`;
  }
}

export default Item;