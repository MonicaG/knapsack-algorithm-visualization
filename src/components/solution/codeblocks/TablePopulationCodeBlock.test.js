import renderer from 'react-test-renderer';
import TablePopulationCodeBlock from './TablePopulationCodeBlock';
import KnapSackAlgorithm from '../../../models/KnapsackAlgorithm';
import Item from '../../../models/Item';

/** 
 * Using snapshot testing because the syntax hightlighting uses a lot of span tags. This makes it difficult to match text as the text 
 * is spread across multiple elements. 
*/

describe('the code block', () => {
  const item1 = new Item('item1', 10, 3);
  const item2 = new Item('item2', 20, 1);
  const item3 = new Item('item3', 25, 2);
  const items = [item1, item2, item3];

  const algorithm = new KnapSackAlgorithm(items, 5);
  const solutionTable = algorithm.solutionTable;

  it('displays the code block when an item fits in the knapsack', () => {
    const capacity = 3
    const tree = renderer
      .create(<TablePopulationCodeBlock
        item={item1}
        capacity={capacity}
        index={1}
        solutionTable={solutionTable} />)
      .toJSON();

    expect(tree).toMatchSnapshot()
  })

  it('displays the code block when an item does NOT fit in the knapsack', () => {
    const capacity = 2
    const tree = renderer
      .create(<TablePopulationCodeBlock
        item={item1}
        capacity={capacity}
        index={1}
        solutionTable={solutionTable} />)
      .toJSON();

    expect(tree).toMatchSnapshot()
  
  });
});