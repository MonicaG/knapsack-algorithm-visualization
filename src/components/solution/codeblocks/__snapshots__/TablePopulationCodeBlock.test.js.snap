// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the code block displays the code block when an item does NOT fit in the knapsack 1`] = `
<pre
  style={
    Object {
      "background": "#f8f8ff",
      "color": "#000",
      "display": "block",
      "overflowX": "auto",
      "padding": "0.5em",
    }
  }
>
  <code
    className="language-javascript"
    style={
      Object {
        "whiteSpace": "pre",
      }
    }
  >
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // w: item weight (3)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        2
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // value: item value (10)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        3
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // c: current capacity (2)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        4
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // i: index (1)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        5
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#954121",
          }
        }
      >
        if
      </span>
      <span
        style={Object {}}
      >
         (w &lt;= c) { 
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // 3 &lt;= 2
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        6
      </span>
      <span
        style={Object {}}
      >
          T[i][c] = 
      </span>
      <span
        style={
          Object {
            "color": "#0086b3",
          }
        }
      >
        Math
      </span>
      <span
        style={Object {}}
      >
        .max(T[i-
      </span>
      <span
        style={
          Object {
            "color": "#40a070",
          }
        }
      >
        1
      </span>
      <span
        style={Object {}}
      >
        ][c], (value + T[i - 
      </span>
      <span
        style={
          Object {
            "color": "#40a070",
          }
        }
      >
        1
      </span>
      <span
        style={Object {}}
      >
        ][c - w])) 

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        7
      </span>
      <span
        style={Object {}}
      >
        }
      </span>
      <span
        style={
          Object {
            "color": "#954121",
          }
        }
      >
        else
      </span>
      <span
        style={Object {}}
      >
         {

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        8
      </span>
      <span
        style={Object {}}
      >
          T[i][c] = T[i-
      </span>
      <span
        style={
          Object {
            "color": "#40a070",
          }
        }
      >
        1
      </span>
      <span
        style={Object {}}
      >
        ][c] 

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        9
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        /* 

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "color": "#408080",
          "display": "block",
          "fontStyle": "italic",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        10
      </span>
      T[1][2] = T[0][2]

    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "color": "#408080",
          "display": "block",
          "fontStyle": "italic",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        11
      </span>
      T[1][2] = 0

    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        12
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        */
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        13
      </span>
      }

    </span>
  </code>
</pre>
`;

exports[`the code block displays the code block when an item fits in the knapsack 1`] = `
<pre
  style={
    Object {
      "background": "#f8f8ff",
      "color": "#000",
      "display": "block",
      "overflowX": "auto",
      "padding": "0.5em",
    }
  }
>
  <code
    className="language-javascript"
    style={
      Object {
        "whiteSpace": "pre",
      }
    }
  >
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // w: item weight (3)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        2
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // value: item value (10)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        3
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // c: current capacity (3)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        4
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // i: index (1)
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        5
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#954121",
          }
        }
      >
        if
      </span>
      <span
        style={Object {}}
      >
         (w &lt;= c) { 
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        // 3 &lt;= 3
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        6
      </span>
      <span
        style={Object {}}
      >
          T[i][c] = 
      </span>
      <span
        style={
          Object {
            "color": "#0086b3",
          }
        }
      >
        Math
      </span>
      <span
        style={Object {}}
      >
        .max(T[i-
      </span>
      <span
        style={
          Object {
            "color": "#40a070",
          }
        }
      >
        1
      </span>
      <span
        style={Object {}}
      >
        ][c], (value + T[i - 
      </span>
      <span
        style={
          Object {
            "color": "#40a070",
          }
        }
      >
        1
      </span>
      <span
        style={Object {}}
      >
        ][c - w])) 

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        7
      </span>
      <span
        style={Object {}}
      >
        
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        /* 

      </span>
    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "color": "#408080",
          "display": "block",
          "fontStyle": "italic",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        8
      </span>
      T[1][3] = Math.max(T[0][3], (10 + T[0][0])) 

    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "color": "#408080",
          "display": "block",
          "fontStyle": "italic",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        9
      </span>
      T[1][3] = Math.max(0, 10)

    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "color": "#408080",
          "display": "block",
          "fontStyle": "italic",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        10
      </span>
      T[1][3] = 10

    </span>
    <span
      style={
        Object {
          "backgroundColor": "#BEF264",
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        11
      </span>
      <span
        style={
          Object {
            "color": "#408080",
            "fontStyle": "italic",
          }
        }
      >
        */
      </span>
      <span
        style={Object {}}
      >
        

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        12
      </span>
      <span
        style={Object {}}
      >
        }
      </span>
      <span
        style={
          Object {
            "color": "#954121",
          }
        }
      >
        else
      </span>
      <span
        style={Object {}}
      >
         {

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        13
      </span>
      <span
        style={Object {}}
      >
          T[i][c] = T[i-
      </span>
      <span
        style={
          Object {
            "color": "#40a070",
          }
        }
      >
        1
      </span>
      <span
        style={Object {}}
      >
        ][c] 

      </span>
    </span>
    <span
      style={
        Object {
          "display": "block",
        }
      }
    >
      <span
        className="comment linenumber react-syntax-highlighter-line-number"
        style={
          Object {
            "display": "inline-block",
            "minWidth": "2.25em",
            "paddingRight": "1em",
            "textAlign": "right",
            "userSelect": "none",
          }
        }
      >
        14
      </span>
      }

    </span>
  </code>
</pre>
`;
